### DISCUSSION TITLE
mosquitto_pub status info

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16260] [Q&A] [answered]
I notice that when I use this command to get a status;

mosquitto_pub -h 192.168.1.134 -p 8883 --cafile /etc/ssl/certs/ca-bundle.crt -u "garden" -P "123456" -t cmnd/garden/Status -m 0

... the information is shown on the MQTT server log but not on the console running the command. 
Is there a way to retrieve the information to the console running the command?

**Comment by sfromis** [comment, marked as answer]
You can subscribe to payloads from the device, with whatever tools, the command `mosquitto_sub` can do that and comes in a pair with `mosquitto_pub`. MQTT is a pubsub protocol, meaning that there are no synchronous responses, as commands are not addressed to devices. Instead devices subscribe to what topic(s) they want messages from.

In your case, it may be a bit easier using command `Status0` (and an empty payload) for one long result message instead of `Status 0` with many lines.

Anyway, `mosquitto_pub` and `mosquitto_sub` are mainly debugging tools, and are not really practical as a strategy for doing regular operation.

**Reply by barbudor** [reply]
> Is there a way to retrieve the information to the console running the command?

In general, no because MQTT is meant to be asynchronous and not "connected" so the response, if any, is not specifically addressed to the one who sent the request because the broker doesn't have any way of "connecting" one with the other.
The broker just send the message to anyone subscribing to a topic.
When you publish to "cmnd/mytasmota/status0" the broker just forward that to anyone subscribing to the topic. Actually tasmota on the device has susbscribed to "cmnd/mytasmota/#"
When the device replies, the message is sent to anyone subscribing to the response topic

So it is not 1-to-1 "connected" communication like a http command

In order to use MQTT into some script, I have my own python script which does a "publish and wait response" by actually subscribing to response topic before sending the command.
I attach it here in case it can help but with no support 😄 
[mwait.zip](https://github.com/arendst/Tasmota/files/9381480/mwait.zip)

**Comment by dac2020** [comment]
Thank you for the information. A lot to absorb but working on it :).
Yes, using those commands mainly for debugging while setting things up etc.

I decided to install a home assistant server as well so I can learn how all this comes together.

Some of the things not yet fully clear to me are...

If I understand, MQTT is a protocol that is not specific to tasmota, other devices that communicate using this protocol can use the MQTT server so I'm not limited to tasmota when using MQTT. I can have a mix of hardware, some not using tasmota.

Is this correct?

**Reply by sfromis** [reply]
That is correct. MQTT is a general purpose lightweight pubsub protocol much older than Tasmota (started in 1999), and well-suited for handling sensors and IoT devices, without high bandwidth demands.  https://en.wikipedia.org/wiki/MQTT
To roll your own programming/automation based on MQTT, Node-Red is a good tool, and can be recommended. Home Assistant is much more prebaked system designed to avoid doing programming and protocol handling, while still providing a lot of smarthome-related features.

One reason why I very much like device communication via MQTT is that it makes it very easy to listen to the communication, and do extra things like automations or logging. That is generally much harder with protocols designed for one-to-one closed communication.

**Comment by dac2020** [comment]
Seems I have a lot to learn but slowly starting to.

I put together an MQTT broker on centos and wanted to log the messages to DB for later review and just to learn how.
I found this; https://ncd.io/log-mqtt-gateway-data-to-mysql-database/

... but it turns out it's for their own hardware so it's not working with my MQTT broker so far.
If anyone knows of a way to log to mysql, please share a link.

**Reply by sfromis** [reply]
That would be reasonably easy to put together with Node-Red. Of course, can become involved if you also want the data normalized into a bunch of specific tables.

My current method with Node-Red is even simpler, daily rolling flat files based on simply subscribing to # and logging all. Very flexible usage, and still relatively straightforward to deal with when parsing the JSON payloads.

Anyway, the page you found does not appear to have any ties to their hardware or specific software. It just points to a small piece of Python code, which you have to adapt a little bit for your setup.
https://github.com/ncd-io/Mosquitto_Subscriber_MySQL_Publisher/blob/master/Mosquitto_Subscriber_MySQL_Publisher.py
However, it seems specific to a particular use case, and would need further adaption to be well-suited for data from Tasmota.

**Reply by dac2020** [reply]
Just FYI;

I tried to get it running.

$ python3 mqtt_sub_mysql_pub.py
MQTT Client Connected
MySQL Client Connected
MQTT Client Connected
MySQL Client Connected
MQTT Client Connected
MySQL Client Connected
^CTraceback (most recent call last):
  File "mqtt_sub_mysql_pub.py", line 118, in <module>
    client.loop_forever()
  File "/usr/local/lib/python3.6/site-packages/paho/mqtt/client.py", line 1773, in loop_forever
    self._reconnect_wait()
  File "/usr/local/lib/python3.6/site-packages/paho/mqtt/client.py", line 3612, in _reconnect_wait
    time.sleep(min(remaining, 1))
KeyboardInterrupt

The author says it would take 'some work' to get it working. I actually wanted something that logs in real time rather than daily.
If I find something, I"ll share it here.

**Reply by sfromis** [reply]
May be easier to write yourself, instead of doing too much searching....

**Reply by dac2020** [reply]
Wish I was a programmer, I definitely would :). 
Searching the net, it seems there might be some scripts out there. Once I find something, I'll share or maybe I need to start a new question for others to find.

**Reply by barbudor** [reply]
May be do do something simpler using `mosquitto_sub -t #` sent to a log file and using log-rotate
However I'm not sure that SQL is the best target to store logs
There are other logging tools like ELK (Elasticsearch + Logstash + Kibana), but that's another topic for another place

**Reply by dac2020** [reply]
This is overkill but seems to be what I am interested in but at a much higher level that I don't need.

https://github.com/vernemq/vernemq

**Reply by dac2020** [reply]
My goal is to log only changed statuses into a database so I can always have some historical information for the different sensors and locations I might install devices at. I can already see I'll be installing some at my home, my family, some friends and even at work.
Since most of the people I would help are not very technical, I'd like to have logs to see for myself.

I'm familiar with mysql, that's why I have been looking for something that would work with it.

**Reply by sfromis** [reply]
When you have goals, you need to get a little bit of programming skills to match....

**Reply by dac2020** [reply]
I understand that and wish I could :).
Just cannot do it all and have to use what's out there. The best I can do is learn about this, put it to use, help others to do the same.
